vdl_tools/shared_tools/embedding_network/embedding_network.py

```diff
@@ -428,93 +558,69 @@ def get_cluster_sentences_from_text(
+        )
     return nodesdf
-    # nodesdf["clus_sentence_reviewed"] = nodesdf["Cluster"].map(cluster_id_to_review)
-    # return nodesdf
 
+def get_cluster_kwdnames_from_text(
+        nodesdf,
+        textcol,
+        clusattr,
+        clusname,  # column name for the cluster names
+        n_tags,
+        subject=None,
+        n_entities=20,  # int → use top‑k; None → token‑limit fallback
+        model="gpt-4.1-mini",
+        sample_texts=None,
+        improve=False,
+        improve_model='o3-mini'
 
-def get_cluster_names_from_text(
-    nodesdf,
-    textcol,
-    clusattr,
-    clusname,               # column name for the cluster names
-    n_tags,
-    subject=None,
-    n_entities=20,          # int → use top‑k; None → token‑limit fallback
-    weight_param=None,
-    model="gpt-4.1-mini",
 ):
     """
-    Adds two columns to *nodesdf*:
+    Adds one column to *nodesdf*:
         • clusname : list[str]   (up to n_tags keywords)
-        • clus_sum : str         (one‑sentence summary)
-    ------------
-    • Optional *n_entities* (top‑k) with fallback to add_text_below_token_limit.
 
+    If sample_texts are provided, it uses those; otherwise, it calculates new samples.
     """
+    # Define centrality column directly
+    centrality_col = "ClusterCentrality"
 
-    clus_centrality = "ClusterCentrality"
     assistant_prompt = define_assistant_prompt(subject)
 
-    if weight_param is not None:
-        nodesdf["wtd_cc"] = nodesdf[clus_centrality] * nodesdf[weight_param]
-    else:
-        nodesdf["wtd_cc"] = nodesdf[clus_centrality]
-
-    ids_text_prompts, ids_text_sums = [], []
-
-    for clus, cdf in nodesdf.groupby(clusattr):
-        cdf_sorted = cdf.sort_values("wtd_cc", ascending=False)
+    ids_text_prompts = []
 
-        if n_entities is None:
-            cluster_texts = add_text_below_token_limit(cdf_sorted, textcol, model)
-        else:
-            cluster_texts = cdf_sorted.iloc[:n_entities][textcol].values
+    # Use pre-calculated sample texts if provided, otherwise sample new ones
+    if sample_texts is not None:
+        # If sample_texts are provided, use them directly
+        cluster_samples = {clus: sample_texts[clus] for clus in nodesdf[clusattr].unique()}
+    else:
+        # Get the cluster samples using stratified sampling based on centrality
+        cluster_samples = sample_cluster_texts_by_percentile(
+            nodesdf, textcol, clusattr, max_samples=n_entities
+        )
 
-        ids_text_prompts.append((clus, get_text_for_entities(n_tags, cluster_texts)))
-        ids_text_sums.append((clus, get_text_for_summaries(cluster_texts, subject)))
+    # Iterate over the clusters and prepare the cluster texts
+    for clus, sampled_texts in cluster_samples.items():
+        # Use the sampled texts for each cluster
+        cluster_texts = sampled_texts  # Already sampled texts
+        ids_text_prompts.append((clus, format_kwds_prompt(n_tags, cluster_texts)))
```

Comment:
```suggestion
        ids_text_prompts.append((clus, format_kwds_prompt(n_tags, sampled_texts)))
```
Did I misunderstand why we need to set `cluster_text`?